File: nge-web/src/main/resources/org/ngengine/web/WebBinds.js
 
import Binds from "./WebBindsHub.js";
import ImageLoader from "./ImageLoader.js";
import Nip07 from "./Nip07.js";
// convert various buffer types to Uint8Array
const _u = (data) => {
    if (data instanceof Uint8Array) {
        return data;
    } else if (data instanceof Int8Array) {
        return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else if (Array.isArray(data)) {
        return new Uint8Array(data);
    } else if (data instanceof ArrayBuffer) {
        return new Uint8Array(data);
    } else if (data instanceof Uint8ClampedArray) {
        return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else if (data instanceof DataView) {
        return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else if (typeof Buffer !== 'undefined' && data instanceof Buffer) {
        return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    } else {
        throw new TypeError('Unsupported data type for conversion to Uint8Array');
    }
};

function isWorker() {
    return (typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope);
}

function s() {
    return ((typeof window !== 'undefined' && window) ||
        (typeof globalThis !== 'undefined' && globalThis) ||
        (typeof global !== 'undefined' && global) ||
        (typeof self !== 'undefined' && self));
}

export const decodeImageAsync = (data /*byte[]*/, filename /*str*/ , targetWidth /*int*/, targetHeight /*int*/, res, rej) => { /* {
        data: Uint8Array,
        width: number,
        height: number
    
    }*/
   if(!isWorker()||!filename.toLowerCase().endsWith('.svg')) {
        ImageLoader.decodeImage(_u(data), filename, targetWidth, targetHeight ).then(res).catch(e=>rej( String(e)));  
   } else{
        Binds.fireEvent("decodeImage", _u(data), filename, targetWidth, targetHeight).then(res).catch(e=>rej( String(e)));
   }
}

export const helloBinds = () => {
    console.log("nge.js is loaded!");
    const g = s();
    if (g) {
        console.log("User Agent: " + (g.navigator ? g.navigator.userAgent : "unknown"));
        console.log("Platform: " + (g.navigator ? g.navigator.platform : "unknown"));
    }
}
 


 





export const loadScriptAsync = async (script, res, rej) => {
    const isWorker = (typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope);
    if(isWorker) {
        try {
            console.log("Loading script: " + script + " in worker");
            importScripts(script);
            console.log("Script loaded: " + script);
            res();
        } catch (e) {
            const code = await fetch(script).then(r => r.text());
            eval(code);
            console.log("Script loaded via fetch+eval: " + script);
            res();
        }
    } else {
        console.log("Loading script: " + script);
        const g = s();
        const scriptElement = g.document.createElement("script");
        scriptElement.src = script;
        g.document.head.appendChild(scriptElement);
    
        scriptElement.onload = () => {
            console.log("Script loaded: " + script);
            res();
        }
        scriptElement.onerror = (e) => {
            console.error("Failed to load script: " + script, e);
            rej(String(e));
        }
    }
};




export const setPageTitle = (title) => {
    Binds.fireEvent("setPageTitle", title);
}

export const toggleFullscreen = (v) =>{
    Binds.fireEvent("toggleFullscreen", v);
}

export const togglePointerLock = (v) =>{
    Binds.fireEvent("togglePointerLock", v);
}

 
// export const addEventListener = (event, fun) =>{
//     Binds.addEventListener(event, fun);
// }

// export const removeEventListener = (event, fun) =>{
//     Binds.removeEventListener(event, fun);
// }

export const waitNextFrame = (callback) => {
    const l = () => {
        try{
            Binds.removeEventListener("render", l);
        } catch(e){
            console.warn("Error removing render listener in waitNextFrame", e);
        }
        try{
            callback();
        } catch(e){
            console.error("Error in waitNextFrame callback", e);
        }
    }
    Binds.addEventListener("render", l);
}

// export const fireEventAsync = (event, args, res,rej) => {
//     Binds.fireEvent(event, ...args).then(res).catch(e=>rej(String(e)));
// }

export const getRenderTargetAsync = (res, rej) => {
    Binds.fireEvent("getRenderTarget").then(res).catch(e=>rej(String(e)));
}

export const addResizeRenderTargetListener = (fun)=>{
    Binds.addEventListener("resizeRenderTarget", fun);
}

export const addSwapRenderTargetListener = (fun)=>{
    Binds.addEventListener("swapRenderTarget", fun);
}

export const addInputEventListener = (event, fun) => {
    Binds.addEventListener(event, fun);
}

export const removeInputEventListener = (event, fun) => {
    Binds.removeEventListener(event, fun);
}




// audio
export const addAudioEndListener = (fun) => {
    Binds.addEventListener("audioSourceEnded", fun);
};

export const createAudioContextAsync = (sampleRate, id, res, rej) => {
    Binds.fireEvent("createAudioContext", sampleRate, id).then(res).catch(e=>rej(String(e)));
};

export const freeAudioContext = (id) => {
    Binds.fireEvent("freeAudioContext", id);
};

export const createAudioBufferAsync = (ctxId, id, f32channelData, lengthInSamples, sampleRate, res, rej) => {
    Binds.fireEvent("createAudioBuffer", ctxId, id, f32channelData, lengthInSamples, sampleRate).then(res).catch(e=>rej(String(e)));
};

export const freeAudioBuffer = (ctxId, bufId) => {
    Binds.fireEvent("freeAudioBuffer", ctxId, bufId);
};

export const createAudioSourceAsync = (ctxId, id,  res, rej) => {
    Binds.fireEvent("createAudioSource", ctxId, id).then(res).catch(e=>rej(String(e)));
};

export const freeAudioSource = (ctxId, srcId) => {
    Binds.fireEvent("freeAudioSource", ctxId, srcId);
};

export const setAudioBufferAsync = (ctxId, srcId, bufId, res, rej) => {
    Binds.fireEvent("setAudioBuffer", ctxId, srcId, bufId).then(res).catch(e=>rej(String(e)));
};

export const setAudioPositional = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioPositional", ctxId, srcId, v);
};

export const setContextAudioEnv = (ctxId, i8data) => {
    Binds.fireEvent("setContextAudioEnv", ctxId, i8data);
};

export const setAudioPosition = (ctxId, srcId, x, y, z) => {
    Binds.fireEvent("setAudioPosition", ctxId, srcId, x, y, z);
};

export const setAudioVelocity = (ctxId, srcId, x, y, z) => {
    Binds.fireEvent("setAudioVelocity", ctxId, srcId, x, y, z);
};

export const setAudioMaxDistance = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioMaxDistance", ctxId, srcId, v);
};

export const setAudioRefDistance = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioRefDistance", ctxId, srcId, v);
};

export const setAudioDirection = (ctxId, srcId, x, y, z) => {
    Binds.fireEvent("setAudioDirection", ctxId, srcId, x, y, z);
};

export const setAudioConeInnerAngle = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioConeInnerAngle", ctxId, srcId, v);
};

export const setAudioConeOuterAngle = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioConeOuterAngle", ctxId, srcId, v);
};

export const setAudioConeOuterGain = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioConeOuterGain", ctxId, srcId, v);
};

export const setAudioLoop = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioLoop", ctxId, srcId, v);
};

export const setAudioPitch = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioPitch", ctxId, srcId, v);
};

export const setAudioVolume = (ctxId, srcId, v) => {
    Binds.fireEvent("setAudioVolume", ctxId, srcId, v);
};

export const getAudioPlaybackRateAsync = (ctxId, srcId, res, rej) => {
    Binds.fireEvent("getAudioPlaybackRate", ctxId, srcId).then(res).catch(e=>rej(String(e)));
};

export const playAudioSourceAsync = (ctxId, srcId, res, rej) => {
    Binds.fireEvent("playAudioSource", ctxId, srcId).then(res).catch(e=>rej(String(e)));
};

export const pauseAudioSourceAsync = (ctxId, srcId, res, rej) => {
    Binds.fireEvent("pauseAudioSource", ctxId, srcId).then(res).catch(e=>rej(String(e)));
};

export const stopAudioSourceAsync = (ctxId, srcId, res, rej) => {
    Binds.fireEvent("stopAudioSource", ctxId, srcId).then(res).catch(e=>rej(String(e)));
};

export const setAudioContextListener = (
    ctxId, 
    px, py, pz, 
    dx, dy, dz, 
    ux, uy, uz
) => {
    Binds.fireEvent("setAudioContextListener", 
        ctxId, 
        px, py, pz, 
        0,0,0,
        dx, dy, dz, 
        ux, uy, uz
    );
}

let baseUrl = null;
export const getBaseURLAsync = (res, rej) => {
    if(baseUrl){
        res(baseUrl);
        return;
    }
    Binds.fireEvent("getBaseURL").then( (url) => {
        baseUrl = url;
        res(url);
    }).catch(e=>rej(String(e)));
}

export const connectNip07BackendAsync = (res, rej) => {
    Nip07.inject().then(res).catch(e=>rej(String(e)));
}


export const pingFrontEnd = () => {
    Binds.fireEvent("ping");
}


export const runWithDelay=(f, t) => {
    setTimeout(f, t);
}

File: nge-web/src/main/java/org/ngengine/web/WebBinds.java
package org.ngengine.web;

import org.ngengine.web.context.WebCanvasElement;
import org.ngengine.web.context.WebContext.CanvasResizeHandler;
import org.ngengine.web.context.WebContext.CanvasSwapHandler;
import org.teavm.jso.JSByRef;
import org.teavm.jso.JSClass;
import org.teavm.jso.JSFunctor;
import org.teavm.jso.JSModule;
import org.teavm.jso.JSObject;
import org.teavm.jso.JSTopLevel;
import org.teavm.jso.core.JSArray;
import org.teavm.jso.dom.events.Event;
import org.teavm.jso.dom.events.EventListener;
import org.teavm.jso.dom.html.HTMLCanvasElement;
import org.teavm.jso.function.JSConsumer;
import org.teavm.jso.typedarrays.Float32Array;
import org.teavm.jso.typedarrays.Int8Array;

@JSClass
public class WebBinds implements JSObject {

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void helloBinds();

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void decodeImageAsync( @JSByRef byte[] data, String filename, int targetWidth, int targetHeight, JSConsumer<WebDecodedImage> resolve, JSConsumer<String> reject);

  


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void loadScriptAsync(String script, JSConsumer<String> resolve, JSConsumer<String> reject);


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setPageTitle(String title);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void toggleFullscreen(boolean v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void togglePointerLock(boolean v);

    // @JSTopLevel
    // @JSModule("./org/ngengine/web/WebBinds.js")
    // public static native void addEventListener(String event, Object fun);


        // WebBinds.addEventListener("resizeRenderTarget", new CanvasResizeHandler() {
        //     @Override
        //     public void onResize(int width, int height) {
        //         canvasWidth = width;
        //         canvasHeight = height;
        //         System.out.println("Resize render target: "+width+"x"+height);
        //     }
        // });

        // // WebBinds.addEventListener("swapRenderTarget", (CanvasSwapHandler)(c -> {
        // //     if (this.canvasTarget != c) {
        // //         this.canvasTarget = c;
        // //     }
        // // }));

        // WebBinds.addEventListener("swapRenderTarget", new CanvasSwapHandler() {
        //     @Override
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void addResizeRenderTargetListener(CanvasResizeHandler fun);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void addSwapRenderTargetListener(CanvasSwapHandler fun);


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void addInputEventListener(String event, EventListener fun);
   
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void removeInputEventListener(String event, EventListener fun);


    // @JSTopLevel
    // @JSModule("./org/ngengine/web/WebBinds.js")
    // public static native void removeEventListener(String event, Object fun);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void waitNextFrame(JSConsumer<Void> callback);


    // @JSTopLevel
    // @JSModule("./org/ngengine/web/WebBinds.js")
    // public static native JSObject fireEventAsync(String event, JSArray<JSObject> args, JSConsumer<JSObject> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void getRenderTargetAsync(JSConsumer<WebCanvasElement> resolve, JSConsumer<String> reject);



    // audio binds

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void createAudioContextAsync(int sampleRate, int id, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void freeAudioContext(int ctxId);

    // Audio buffer management
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void createAudioBufferAsync(int ctxId, int id, JSArray<Float32Array> f32channelData, int lengthInSamples, int sampleRate, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void freeAudioBuffer(int ctxId, int bufId);

    // Audio source management
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void createAudioSourceAsync(int ctxId, int id, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void freeAudioSource(int ctxId, int srcId);

    // Source property setters
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioBufferAsync(int ctxId, int srcId, int bufId, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioPositional(int ctxId, int srcId, boolean v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setContextAudioEnv(int ctxId, Int8Array i8data);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioPosition(int ctxId, int srcId, float x, float y, float z);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioVelocity(int ctxId, int srcId, float x, float y, float z);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioMaxDistance(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioRefDistance(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioDirection(int ctxId, int srcId, float x, float y, float z);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioConeInnerAngle(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioConeOuterAngle(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioConeOuterGain(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioLoop(int ctxId, int srcId, boolean v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioPitch(int ctxId, int srcId, float v);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioVolume(int ctxId, int srcId, float v);

    // Playback control
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void playAudioSourceAsync(int ctxId, int srcId, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void pauseAudioSourceAsync(int ctxId, int srcId, JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void stopAudioSourceAsync(int ctxId, int srcId, JSConsumer<Void> resolve, JSConsumer<String> reject);

    // Getters
    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void getAudioPlaybackRateAsync(int ctxId, int srcId, JSConsumer<Float> resolve, JSConsumer<String> reject);


    @JSFunctor
    public static interface AudioEndEvent extends JSObject {
        void onAudioEnd(int ctxId, int srcId);
    }

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void addAudioEndListener(AudioEndEvent fun);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void setAudioContextListener(
        int ctxId, 
        float px, float py, float pz, 
        float dx, float dy, float dz, 
        float ux, float uy, float uz
    );

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void getBaseURLAsync(JSConsumer<String> resolve, JSConsumer<String> reject);


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void connectNip07BackendAsync(JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void connectWebRTCBackendAsync(JSConsumer<Void> resolve, JSConsumer<String> reject);

    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void pingFrontEnd();


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void runWithDelay(JSConsumer<Void> callback, int msDelay);


    @JSTopLevel
    @JSModule("./org/ngengine/web/WebBinds.js")
    public static native void connectClipboardBackendAsync(JSConsumer<Void> resolve, JSConsumer<String> reject);
}

